data/campaigns/fri05.wmf/scripting/mission_thread.lua

```diff
@@ -0,0 +1,309 @@
+function count_in_warehouses(ware)
+   local whs = array_combine(
+      p1:get_buildings("frisians_headquarters"),
+      p1:get_buildings("frisians_warehouse"),
+      p1:get_buildings("frisians_port")
+   )
+   local rv = 0
+   for idx,wh in ipairs(whs) do
+      rv = rv + wh:get_wares(ware)
+   end
+   return rv
+end
+
+function check_fish()
+   -- Slowly replenish fish in fishable nodes where the fish is not entirely depleted.
+   while true do
+      for y=0,map.height-1 do
+         for x=0,map.width-1 do
+            local f = map:get_field(x, y)
+            if f.resource == "resource_fish" and f.resource_amount > 0 and f.resource_amount < f.initial_resource_amount then
+               f.resource_amount = f.resource_amount + 1
+            end
+         end
+         sleep(75)
+      end
+   end
+end
+
+function check_trade()
+   local trade_count = 0
+   while true do
+      for _,port in ipairs(p1:get_buildings("frisians_port")) do
+         for _,proposal in ipairs(trade) do
+            local trade_interesting = false
+            for output,amount in pairs(proposal[2]) do
+               if port.flag.ware_economy:needs(output) then
+                  trade_interesting = true
+                  break
+               end
+            end
+            if trade_interesting then
+               for input,amount in pairs(proposal[1]) do
+                  if port:get_wares(input) < amount then
+                     trade_interesting = false
+                     break
+                  end
+               end
+               if trade_interesting then
+                  for input,amount in pairs(proposal[1]) do
+                     port:set_wares(input, port:get_wares(input) - amount)
+                  end
+                  for output,amount in pairs(proposal[2]) do
+                     port:set_wares(output, port:get_wares(output) + amount)
+                  end
+                  trade_count = trade_count + 1
+                  if trade_count == 3 then
+                     local prior_center = scroll_to_field(map:get_field(90, 132))
+                     sleep(1000)
+                     campaign_message_box(training)
+                     scroll_to_map_pixel(prior_center)
+                     prior_center = nil
+                  end
+                  sleep(3600)
+               end
+            end
+         end
+      end
+      sleep(7200)
+   end
+end
+
+function check_objective_wood()
+   local o = add_campaign_objective(obj_wood)
+   while not check_for_buildings(p1, {frisians_woodcutters_house = 3, frisians_foresters_house = 4}) do
+      sleep(4273)
+   end
+   set_objective_done(o)
+   campaign_message_box(secured_wood)
+   while count_in_warehouses("log") < 80 do
```

Comment:
```suggestion
   while count_in_warehouses("log") < 40 do
```
if log is used wisely it does not pile up to 80 very early so the message is triggered very late. but this was not intended.